lightning/src/ln/channelmanager.rs

```diff
@@ -2812,11 +2835,11 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 					self.fail_htlc_backwards_internal(channel_state,
 						htlc_src, &payment_hash, HTLCFailReason::Reason { failure_code, data: onion_failure_data});
 				},
-				HTLCSource::OutboundRoute { session_priv, .. } => {
+				HTLCSource::OutboundRoute { session_priv, mpp_id, .. } => {
 					if {
 						let mut session_priv_bytes = [0; 32];
 						session_priv_bytes.copy_from_slice(&session_priv[..]);
-						self.pending_outbound_payments.lock().unwrap().remove(&session_priv_bytes)
+						self.pending_outbound_payments.lock().unwrap().remove(&(session_priv_bytes, mpp_id))
```

Comment:
So it looks like we could still get a `PaymentFailed` followed by a `PaymentSent` event, or possibly multiple `PaymentFailed` events here? Its somewhat unclear exactly what the solution should be, though - ideally we'd like to wait and only call it a "PaymentFailed" failure if all MPP fragments fail, but at the same time users need to be notified if a fragment fails so that they can retry that one path. Maybe we need a new `MPPFragmentFailed` event and then only generate `PaymentFailed` if *all* fragments fail (ie there are no more with the same mpp id)?